pkg/common/consumer/consumer_handler.go

```diff
@@ -113,14 +124,25 @@ func (consumer *SaramaConsumerHandler) ConsumeClaim(session sarama.ConsumerGroup
 	// The `ConsumeClaim` itself is called within a goroutine, see:
 	// https://github.com/Shopify/sarama/blob/master/consumer_group.go#L27-L29
 	for message := range claim.Messages() {
-
 		if ce := consumer.logger.Desugar().Check(zap.DebugLevel, "debugging"); ce != nil {
 			consumer.logger.Debugw("Message claimed", zap.String("topic", message.Topic), zap.Binary("value", message.Value))
 		}
 
-		// Don't use the session context since it is closed before messages are drained.
-		// Handle must finish before the session timeout.
-		mustMark, err := consumer.handler.Handle(context.Background(), message)
+		// Preemptively interrupt processing messages if the session is closed.
+		// Processing all messages from the buffered channel can take a long time,
+		// potentially exceeding the session timeout, leading to duplicate events.
+		if sessionClosed(session) {
+			consumer.logger.Infof("Session closed for %s/%d. Exiting ConsumeClaim ", claim.Topic(), claim.Partition())
+			break
+		}
+
+		// Don't use the session context because when cancelled it will interrupt the in-flight request,
+		// potentially leading to duplicates
+
+		// Handle must finish before the session timeout. Enforcing it on the client side.
+		ctx, cancel := context.WithTimeout(context.Background(), consumer.timeout)
```

Comment:
What will happen when the delivery retry spec is longer than the timeout? I feel this this will override it .

Reply:
should that perhaps than be included in the math for https://github.com/knative-sandbox/eventing-kafka/pull/659/files#diff-318b63710b5768647bfeb5b9bca7a04b80f3a8388439a6b8bc50eddfb8fed8e7R69 ?

Reply:
KafkaSource does not retry... so yes seems like it's not the right place to do that